src/Microsoft.AspNet.Razor/SourceLocation.cs

```diff
@@ -45,43 +86,78 @@ public override string ToString()
                 CharacterIndex);
         }
 
+        /// <inheritdoc />
         public override bool Equals(object obj)
         {
             return (obj is SourceLocation) && Equals((SourceLocation)obj);
         }
 
+        /// <inheritdoc />
         public override int GetHashCode()
         {
             // LineIndex and CharacterIndex can be calculated from AbsoluteIndex and the document content.
-            return AbsoluteIndex;
+            return HashCodeCombiner.Start()
+                .Add(FilePath, StringComparer.Ordinal)
+                .Add(AbsoluteIndex)
+                .CombinedHash;
         }
 
+        /// <inheritdoc />
         public bool Equals(SourceLocation other)
         {
-            return 
+            return string.Equals(FilePath, other.FilePath, StringComparison.Ordinal) &&
                 AbsoluteIndex == other.AbsoluteIndex &&
                 LineIndex == other.LineIndex &&
                 CharacterIndex == other.CharacterIndex;
         }
 
+        /// <inheritdoc />
         public int CompareTo(SourceLocation other)
         {
+            var filePathOrdinal = string.Compare(FilePath, other.FilePath, StringComparison.Ordinal);
+            if (filePathOrdinal != 0)
+            {
+                return filePathOrdinal;
+            }
+
             return AbsoluteIndex.CompareTo(other.AbsoluteIndex);
         }
 
-        public static SourceLocation Advance(SourceLocation left, string text)
+        /// <summary>
+        /// Advances the <see cref="SourceLocation"/> by the length of the <paramref name="text" />.
+        /// </summary>
+        /// <param name="left">The <see cref="SourceLocation"/> to advance.</param>
+        /// <param name="text">The <see cref="string"/> to advance <paramref name="left"/> by.</param>
+        /// <returns>The advanced <see cref="SourceLocation"/>.</returns>
+        public static SourceLocation Advance(SourceLocation left, [NotNull] string text)
         {
             var tracker = new SourceLocationTracker(left);
             tracker.UpdateLocation(text);
             return tracker.CurrentLocation;
         }
 
-        public static SourceLocation Add(SourceLocation left, SourceLocation right)
+        /// <summary>
+        /// Adds two <see cref="SourceLocation"/>s.
+        /// </summary>
+        /// <param name="left">The left operand.</param>
+        /// <param name="right">The right operand.</param>
+        /// <returns>A <see cref="SourceLocation"/> that is the sum of the left and right operands.</returns>
+        /// <exception cref="ArgumentException">if the <see cref="FilePath"/> of the left and right operands
+        /// are different.</exception>
+        public static SourceLocation operator +(SourceLocation left, SourceLocation right)
         {
+            if (!string.Equals(left.FilePath, right.FilePath, StringComparison.Ordinal))
+            {
+                throw new ArgumentException(
```

Comment:
did you say `+` and `-` are rarely used?  wondering because `(A - B) + C` will always `throw` now.  this is why I thought we should allow one `FilePath` to be `null` even if the other isn't.